src/control/cmd/daos_server/storage_nvme.go

```diff
@@ -0,0 +1,250 @@
+//
+// (C) Copyright 2022 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package main
+
+import (
+	"os/user"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/cmd/dmg/pretty"
+	"github.com/daos-stack/daos/src/control/common/cmdutil"
+	"github.com/daos-stack/daos/src/control/logging"
+	"github.com/daos-stack/daos/src/control/server"
+	"github.com/daos-stack/daos/src/control/server/config"
+	"github.com/daos-stack/daos/src/control/server/storage"
+)
+
+type nvmePrepareResetFn func(storage.BdevPrepareRequest) (*storage.BdevPrepareResponse, error)
+
+type nvmeCmd struct {
+	Prepare prepareDrivesCmd `command:"prepare-drives" description:"Prepare NVMe SSDs for use by DAOS"`
+	Release releaseDrivesCmd `command:"release-drives" description:"Release NVMe SSDs for use by OS"`
+	Scan    scanDrivesCmd    `command:"scan" description:"Scan NVMe SSDs"`
+}
+
+func getTargetUser(reqUser string) (string, error) {
+	if reqUser == "" {
+		runningUser, err := user.Current()
+		if err != nil {
+			return "", errors.Wrap(err, "couldn't lookup running user")
+		}
+		if runningUser.Username == "" {
+			return "", errors.New("empty username returned from current user lookup")
+		}
+		return runningUser.Username, nil
+	}
+	return reqUser, nil
+}
+
+func validateVFIOSetting(targetUser string, reqDisableVFIO bool, iommuEnabled bool) error {
+	if targetUser != "root" {
+		if reqDisableVFIO {
+			return storage.FaultBdevNonRootVFIODisable
+		}
+		if !iommuEnabled {
+			return storage.FaultBdevNoIOMMU
+		}
+	}
+	return nil
+}
+
+func updatePrepReqParams(log logging.Logger, iommuEnabled bool, req *storage.BdevPrepareRequest) error {
+	targetUser, err := getTargetUser(req.TargetUser)
+	if err != nil {
+		return err
+	}
+	// Update target user parameter in request.
+	req.TargetUser = targetUser
+
+	if err := validateVFIOSetting(targetUser, req.DisableVFIO, iommuEnabled); err != nil {
+		return err
+	}
+
+	switch {
+	case req.DisableVFIO:
+		log.Info("VMD not enabled because VFIO disabled in command options")
+	case !iommuEnabled:
+		log.Info("VMD not enabled because IOMMU disabled on platform")
+	default:
+		// If none of the cases above match, set enable VMD flag in request.
+		req.EnableVMD = true
+	}
+
+	return nil
+}
+
+type prepareDrivesCmd struct {
+	cmdutil.LogCmd  `json:"-"`
+	helperLogCmd    `json:"-"`
+	iommuCheckerCmd `json:"-"`
+
+	PCIBlockList string `long:"pci-block-list" description:"Whitespace separated list of PCI devices (by address) to be ignored when unbinding devices from Kernel driver to be used with SPDK (default is no PCI devices)"`
```

Comment:
Does this actually work on the command line with whitespace separation, or are quote marks required around the list?

Reply:
quote marks are required

Reply:
address list input parameters now require comma-separation, done